### ISSUE TITLE
Error adding "filters" array to an object's ._stateProperties using object.saveState

### ISSUE THREAD

**Issue by sjscotti** [fabricjs/fabric.js#6452]
## Version
3.6.3

## Test Case
https://codesandbox.io/s/gracious-carson-l5ueh

## Information about environment
browser? Firefox 
Which browsers?

## Steps to reproduce  -- see the fiddle shown above.
I have an image object `image` and I add a simple greyscale filter to it.   I want to save the filter into the image object's `_stateProperties`.
I use the following call to add a user-defined entry into `_stateProperties`

` image.saveState({ stateProperties: ["filters"] });`

When I do a console.log on `image`, I can see the correct entry for the filter constructor in the array for `filters` in the image object, but when I look at `filters` in the image object's  `_stateProperties`, there is no constructor in the array but an unknown object is showing up.

## Expected Behavior
from console.log:  I expected to see the same constructor in the array that is shown in `.filters` to be in `_stateProperties.filters`

## Actual Behavior
I can only see an object as describe above in the console.log.  BTW, the greyscale image is displayed in the browser window correctly.

**Comment by asturur** [comment]
What do you mean by constructor? the saveState method creates simple javascript objects.
Filters are classes and saveState was meant to save some properties of the fabricObjects, i m not sure is sane to have keep track of the filters too.
statefull part of fabricJS is not great and i m not sure it should persist forever.

What are you using it for?

**Comment by sjscotti** [comment]
Sorry for my poor wording.  I am using fabric.js as the back end for an online drawing program that kids can use.  I have been doing undo/redo by saving the whole canvas at each step based on the fabric-history repository (https://github.com/lyzerk/fabric-history), but there are alot of images in the palette that kids can add and colorize by using filters.  With all the images, the size of the JSON of the whole canvas being saved at each step is hurting performance after several dozen steps, so I want to develop an undo/redo that works more directly with the objects and not the whole canvas.  

With that long-winded explanation of what I am trying to do.  I thought that I could use the existing `._stateProperties` to hold the info of the prior state of an object for an undo, but I needed to also have the `.filters` info included because a step to be saved may be to just colorize an image object.  So the `.filters` I want to store in `._stateProperties` would need to be identical to the entry in the object's `.filters` at the prior state.  I thought that was supposed to be possible with the `image.saveState({ stateProperties: ["filters"] });` command I used, and when it didn't work, I assumed it was a bug.  Is there another way I should have called `image.saveState` to save the `.filters`, or is saveState not intended for the purpose I have described? (and if not, is there another approach I should use?)
Thanks!

**Comment by asturur** [comment]
Sorry for the long delay.
I suggest to not use ._stateProperties for undo, is just no the right tool for the job.

Undo / Redo in an appliacation that needs to go back in time with the state of the fabric canvas usually uses full JSON dumps in an array and uses loadFromJSON to restore older points.

Then from there you can optimize minimizing resource reloading and other things, but really that should be the path to follow.

**Comment by sjscotti** [comment]
Thanks for getting back with me.   I had a full save/restore of the canvas from JSON dumps that, as I mentioned above, was very resource intensive.  So I started implementing a command pattern approach based on what Mr. Nikitsenka described in his answer to https://stackoverflow.com/questions/61459711/fabricjs-implementing-undo-redo-on-canvases/61496587#61496587 and in his answer to my question https://stackoverflow.com/questions/62576458/obtaining-initial-state-of-target-object-in-objectmodified-event-in-fabric-js    So I have gotten pretty far along in implementing undo that is based on the `._stateProperties` capability and I would like to stay with that approach.  I found that I could add the `.filters` key very simply to `._stateProperties` after a standard call (i.e., with no arguments) to `object.saveState()`.  But the issue that I am having trouble with is getting a clone of the `.filters` object to associate with the new `filters` key in `._stateProperties`, since by default `.filters` is passed by reference.  I tried a JSON approach to clone an  object `.filters`, but it must leave out some elements that is needed by fabric.js.  If `object.saveState()` using an argument to add `filters` property doesn't work to clone the `filters` of an object into `._stateProperties`, is there another way I can clone `filters` so it will keep all the original elements of `filters`?
Thanks!

**Comment by asturur** [comment]
The point is that saveState is made for key values pairs that contains primitives and not classes.
Is made to compare if something is changed and not to restore it, so is basically flawled.

Can you move out your implementation keeping a similar approach, but having your own `saveStateBetter`  ( that you can add to the Object prototype ) and that instead of cloning the object properties, create a json dump of the object? In this case filters are cloned and can be restored easily.

I get how cloning all the canvas each step is expensive with many objects.
But saveState approach with nested objects has references all over the place.

In that transform command that i briefly read, there should be a place for using  fabricObject.toObject(), eventually with omitting  defaults to minimize the number of properties saved each time.

If we want to move forward, please make a small working fiddle with your undo/redo implementation.

**Comment by sjscotti** [comment]
I appreciate the suggestions for an improved, general approach to saving states that would include an object's `.filters` array that is not easily cloned.  In looking at my code I think I have a workaround that handles the image filters as a special case and that eliminates the need to clone `.filters` for implementing my undo capability.  I realized that when I change the colors of an image object using new`BlendColor` and/or `BlendImage`, I had just redefined the object's `.filters` array to zero length before pushing the new filters into the `.filter` array.  This approach just 'threw away' the old `.filters` array - the one I wanted to clone.  So if I saved the old `.filters` array instead, I could later add it to my  objects's `_stateProperties` (the object created by the `object.saveState()` method) and use it in my undo approach.  After a bit of trial and error on when to add the `filters` to my saved state without it later being overwritten, and adding additional checks for this special case, it appears to do what I want.   The approach seems to work for single objects and for a set of objects in an active selection.  I am sure I will need to do additional checking (especially if some of the objects are groups), but for now I think this is the easiest approach for my purposes.  Best regards, and thanks for developing such a wonderful library!
